### ISSUE TITLE
[3.0x] GUI ColorPickerButton Events needed

### ISSUE THREAD

**Issue by RaXaR** [godotengine/godot#13456]
**Issue description:**
I need to check if the ColorPicker is open or not. However, I don't see any obvious way to do so.
Would it be possible to add a 'picker_opened' and 'picker_closed' events on it (next to 'color_change')?

Thanks

**Comment by Noshyaar** [comment]
I think the workaround is to use `visibility_changed` signal or `NOTIFICATION_VISIBILITY_CHANGED` notification, or put it inside a Popup.

**Comment by RaXaR** [comment]
I've sort of come to the same conclusion, that I'll need to create a separate scene or popup for this.

Also, I have been trying to edit ```/scene/gui/color_picker.cpp``` to add these events.
So far I've had no luck in triggering the closed signal. It seems that there are two ``` popup->hide()``` calls, but I think due to their scope I cannot do emit_signal("picker_closed") near them. 

I'll continue to tinker a bit more before attemping the workaround.

Thanks

**Comment by RaXaR** [comment]
Phew, Ok I got it to emit the signals as I wanted and can now action them in GDScript callbacks.

However, in the console I see this now:
```
ERROR: Nonexistent signal: picker_closed
   At: core\object.cpp:1460
ERROR: Nonexistent signal: picker_opened
   At: core\object.cpp:1460
```
@Noshyaar I've added these two signals to the ```/scenes/gui/color_picker.cpp``` file. I'm not sure if I need to add them anywhere else?

**Comment by Noshyaar** [comment]
Hard to say without seeing the changes. I guess `ADD_SIGNAL` in bind_method?

**Comment by RaXaR** [comment]
Yip, I've added it in the bind methods, here's a Gist of the [the .cpp and .h](https://gist.github.com/RaXaR/e69ee91163670545d3d07942b5c58c08) files.

I'm wondering if I need to add it anywhere else in the code, as in a registry somewhere, hmm

**Comment by RaXaR** [comment]
Aha, it's when a 'Disconnect' the signal via the Godot UI. Then those errors pop up.

**Comment by MarianoGnu** [comment]
are you using the color picker directly or a color picker button?

- For first case: object.visible should work
http://docs.godotengine.org/en/latest/classes/class_canvasitem.html#member-variables

for second case: object.get_picker().visible should work
http://docs.godotengine.org/en/latest/classes/class_colorpickerbutton.html?highlight=color%20picker#class-colorpickerbutton-get-picker

**Comment by Noshyaar** [comment]
@MarianoGnu It's actually a ColorPickerButton.

You can expose `popup` (PopupPanel) to GDScript and connect the Popup signals.

**Comment by RaXaR** [comment]
@MarianoGnu, @Noshyaar I tried to connect() the ColorPicker popup via the get_picker() method.
It doens't work. The console says:
```
ERROR: In Object of type 'ColorPicker': Attempt to connect nonexistent signal 'a
bout_to_show' to method 'Control._on_CPB_Popup_about_to_show'
   At: core\object.cpp:1403
ERROR: In Object of type 'ColorPicker': Attempt to connect nonexistent signal 'p
opup_hide' to method 'Control._on_CPB_Popup_hide'
   At: core\object.cpp:1403
```
(I compiled vanilla codebase).

**Comment by Noshyaar** [comment]
`get_picker` returns ColorPicker, not PopupPanel. There's currently no method that exposes the popup. I guess you have to expose it from C++ to GDScript.

**Comment by RaXaR** [comment]
Yeah, but doesn't it make more sense to add these two 'picker_xyz' signals?
I mean they could be reused in other Pickers, like DatePicker, TexturePicker, IconPicker, GrandmaPicker etc? :)

**Comment by Noshyaar** [comment]
Since we don't have Picker class yet, closest thing I could think of is Popup, which already has those signals.

**Comment by RaXaR** [comment]
Understood, I'll add a 'get_popup()' method to ColorPickerButton, see if that works out okay and report back.

**Comment by RaXaR** [comment]
I tested it with
```
	var cp = $ColorPickerButton.get_popup()
	cp.connect("about_to_show", self, "_on_CPB_Popup_about_to_show")
	cp.connect("popup_hide", self, "_on_CPB_Popup_hide")
```
and it worked just fine. :)

**Comment by akien-mga** [comment]
Fixed by #13491.
